be/src/exec/pipeline/exchange/sink_buffer.h

```diff
@@ -44,46 +48,64 @@ class SinkBuffer {
                 auto* closure = new CallBackClosure<PTransmitChunkResult>();
                 closure->ref();
                 closure->addFailedHandler([this]() noexcept {
-                    _in_flight_rpc_num--;
-                    _is_cancelled = true;
+                    _in_flight_rpc_num.fetch_sub(1, std::memory_order_release);
+                    {
+                        std::lock_guard<std::mutex> l(_mutex);
+                        _is_finishing = true;
+                    }
                     LOG(WARNING) << " transmit chunk rpc failed";
                 });
                 closure->addSuccessHandler([this](const PTransmitChunkResult& result) noexcept {
-                    _in_flight_rpc_num--;
+                    _in_flight_rpc_num.fetch_sub(1, std::memory_order_release);
                     Status status(result.status());
                     if (!status.ok()) {
-                        _is_cancelled = true;
+                        {
+                            std::lock_guard<std::mutex> l(_mutex);
+                            _is_finishing = true;
+                        }
                         LOG(WARNING) << " transmit chunk rpc failed, " << status.message();
                     }
                 });
                 _closures[dest_instance_id] = closure;
-
                 _buffers[dest_instance_id] = std::queue<TransmitChunkInfo>();
+                _request_seqs[dest_instance_id] = 0;
             }
-        }
 
-        try {
-            _thread = std::thread{&SinkBuffer::process, this};
-        } catch (const std::exception& exp) {
-            LOG(FATAL) << "[ExchangeSinkOperator] create thread: " << exp.what();
-        } catch (...) {
-            LOG(FATAL) << "[ExchangeSinkOperator] create thread: unknown";
+            _expected_eos = 0;
+            for (auto& [_, num] : _num_sinkers_per_dest_instance) {
+                _expected_eos += num;
+            }
```

Comment:
if for every dest fragment instance, the number of the sinkers are the same, suggest to
use _expect_eos = _num_sinkers_per_dest_instance.size() * _num_sinkers

Reply:
May not equal